File: python/binary_search.py
def find(search_list, value):
    if search_list == []:
        raise ValueError('value not in array')
    l = 0
    r = len(search_list) - 1
    while l <= r:
        mid = (l + r) // 2
        if search_list[mid] == value:
            return mid
        if search_list[mid] < value:
            l = mid + 1
        else:
            r = mid - 1
    raise ValueError('value not in array')

File: ocaml/binary_search.ml
open Base

let find arr target =
  let rec find' i j =
    if i > j then Error "value not in array"
    else
      let mid = i + ((j - i) / 2) in
      if arr.(mid) = target then Ok mid
      else if arr.(mid) < target then find' (mid + 1) j
      else find' i (mid - 1)
  in
  find' 0 @@ (Array.length arr - 1)
